### ISSUE TITLE
[LibOS] Allow ipc worker thread of leader process to exit after the other processes

### ISSUE THREAD

**Issue by llly** [gramineproject/gramine#1168]
### Description of the feature

Add a new manifest libos option that keep ipc worker thread of leader process working, until all forked processes exit.

### Why Gramine should implement it?

Each Gramine process starts an ipc worker thread to manage `vmid` , `id range`, `posix_lock`. The first process is "leader process". The other processes created by `fork()` communicate with leader process to query, request and release `vmid`, `id range`, `posix_lock`.
Aftaer leader process exited, another process cannot successfully execute a syscall that send ipc message to leader process to query or request `vmid` , `id range`, `posix_lock`. For example, `fork()`, `clone()` and `fcntl(F_SETLK)`.
In practice, I see Java workload fork and execve Python in parallel, but Java exit without wait Python. Then Python fails to create new thread.
Double fork fails too. we can run libc function `daemon(1, 1)` and get error log
```
[P2:T2:helloworld] error: Cound not allocate new vmid!
[P2:T2:helloworld] error: Sending IPC process-exit notification failed: Connection reset by peer (ECONNRESET)
```
Now when EDMM is enabled, it's feasible to keep internal thread live and block enclave destroy because it doesn't request too much EPC.

**Comment by dimakuv** [comment]
Yes, this is a known limitation, and we have a TODO about this in the code: https://github.com/gramineproject/gramine/blob/5fb315426a3bcc306969d5ace203513265844c38/libos/src/sys/libos_exit.c#L21-L25

We have the building blocks to implement this, I think:
- https://github.com/gramineproject/gramine/blob/4f8d7446d25711a4983cdf6695f44e4c8a8e01d9/libos/include/libos_process.h#L55-L56
- https://github.com/gramineproject/gramine/blob/0119fd23ebbc892b33868e9ea9296ba6e8061685/libos/src/sys/libos_wait.c#L102

So I think we could have smth like this (not real code, more like pseudo-code to illustrate the idea):
```c
if (!g_process.ppid) {
    /* We are the leader process, do not terminate until all (known) children are terminated. */
    lock(&g_process.children_lock);
    while (!LISTP_EMPTY(&g_process.children)) {
        unlock(&g_process.children_lock);
        int ret = do_waitid(P_ALL, /*id ignored*/0, /*infop=*/NULL, WEXITED);
        if (ret != 0 && ret != -ECHILD && ret != -ERESTARTSYS) {
            /* something went terribly wrong, just die */
            log_error("Fatal error while leader process waiting for other processes to terminate");
            PalProcessExit(1);
        }
        lock(&g_process.children_lock);
    }
    unlock(&g_process.children_lock);
}
```

I think there can be one problem with this approach -- if we have three processes (the leader, the child, the grandchild), and the child terminates but the grandchild still lives, the leader process does **not** know anything about the grandchild and will still terminate itself. Thus the grandchild will again be in the situation explained by the topic starter. On the other hand, we're at least improving *something* in the current situation (and maybe the described corner case doesn't happen in practice).

**Comment by dimakuv** [comment]
Actually, maybe @boryspoplawski could say if my proposed solution makes sense, or it's completely broken?

**Comment by boryspoplawski** [comment]
Before you even tackle this problem, you need to implement reparenting in Gramine - basically the scenario that @dimakuv described with a grandchild process. This is not really trivial, because some apps might not be ready to handle SIGCHLD from processes they did not spawn. On normal Linux you would have an init process (pid 1), which would manage all of that. Maybe it's time to move Gramine to similar approach (given EDMM would limit resource usage for this additional process)? Just thinking out loud.

**Comment by dimakuv** [comment]
This is a duplicate of https://github.com/gramineproject/gramine/issues/21. Thus I'm closing it.
